clang-tidy: Enable misc-no-recursion (#29690)
bitcoin/bitcoin: doc/developer-notes.md

    Use `reinterpret_cast` and `const_cast` as appropriate.
  - Prefer [`list initialization ({})`](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-list) where possible.
    For example `int x{0};` instead of `int x = 0;` or `int x(0);`

>   - Recursive functions are generally discouraged, and checked for by

Reviewer: I think "generally" is too strong and also a bit of a contradiction with the next sentence. How about something like "Recursive functions are discouraged unless they are the best approach for a given problem. In this case  `NOLINTNEXTLINE(misc-no-recursion)` can be used to suppress the `clang-tidy` check."

Reviewer 2: I do not think recursive functions should be "generally discouraged" or "discouraged". Consider this wording instead:

"Think twice before introducing recursive functions. Consider an iterative approach as well. When recursiveness is the best approach, use `NOLINTNEXTLINE(misc-no-recursion)` to suppress the check."

Author: > "Think twice before introducing recursive functions. Consider an iterative approach as well."

This just reads like a long form of "recursion is discouraged" to me.

Reviewer 3: > I think "generally" is too strong

"generally" makes it less strong, since without it the sentence would imply it's always discouraged.

It's getting a bit too bikesheddy for me so I'm going to leave this conversation here. Imo all three suggestions so far pretty much achieve the same intent (and I'd be okay with): recursion should be used if it's clearly the best choice, otherwise don't.

Reviewer: > "generally" makes it less strong, since without it the sentence would imply it's always discouraged.

I disagree and I did not suggest to just strike it but to use a different wording that is more clear. As it is now, I read this as "When you think you want to use recursion, you are generally wrong.". Maybe it's my ego but I don't like that statement ;) I suppose you read this as "When you want to iterate over anything, generally you shouldn't use recursion." and that is also a reasonable interpretation but I don't think the distinction is clear.

>   - Recursive functions are generally discouraged, and checked for by
>     clang-tidy. When recursiveness is the best approach, use

Reviewer 4: Not sure if the point is to discourage recursion, or to tell developers that it has to be marked explicitly? I think we can consider it implied that people should use recursion only when appropriate, just like any other practices that can lead to bugs. And disagreements about whether or not [fill in the blank] is appropriate can be resolved in review on a case-by-case basis.
```suggestion
  - Recursion is checked by clang-tidy and thus must be made explicit. Use
```

Reviewer: That framing makes a lot more sense to me 👍
